File: apps/api/aqapi/core/domain/base_model.py
from abc import ABC, abstractmethod
from datetime import datetime
from typing import Generic, Optional, TypeVar
from aqapi.core.domain.value_object.version import Version
from aqapi.core.domain.value_object.base_id import BaseId
from aqapi.core.entity.base_entity import BaseEntity
from aqapi.family_member.domain.value_object.family_member_id import FamilyMemberId
from aqapi.shared.domain.value_object.screen_id import ScreenId
from aqapi.core.domain.value_object.relation_validator import RelationValidator

EntityType = TypeVar("EntityType", bound='BaseEntity')
IdType = TypeVar("IdType", bound=BaseId)

class BaseModel(ABC, Generic[IdType, EntityType], ):
    """ドメインモデルの基底クラス"""
    
    def __init__(self, 
                 id: IdType, version: Version, 
                 created_at: Optional[datetime] = None, created_by: Optional[FamilyMemberId] = None, created_from: Optional['ScreenId'] = None,
                 updated_at: Optional[datetime] = None, updated_by: Optional[FamilyMemberId] = None, updated_from: Optional['ScreenId'] = None):
        self._id = id
        self._version = version
        self._created_at = created_at
        self._created_by = created_by
        self._created_from = created_from
        self._updated_at = updated_at
        self._updated_by = updated_by
        self._updated_from = updated_from
        self._is_updated = False
        self._relation_validator = RelationValidator()
        self._validate()

    @abstractmethod
    def _validate(self) -> None:
        """モデルの値を検証する"""
        raise NotImplementedError("Subclasses must implement this method to validate model values.")

    @abstractmethod
    @classmethod
    def from_entity(cls, entity: EntityType) -> 'BaseModel':
        raise NotImplementedError("Subclasses must implement this method")

    @property
    def id(self) -> IdType:
        return self._id

    def version(self) -> Version:
        return self._version

    def _update_version(self) -> None:
        """バージョンを1上げ、更新フラグを設定する"""
        if not self._is_updated:
            self._is_updated = True
            self._version = self._version.get_next_version()
    
    def is_same_version(self, other: 'BaseModel') -> bool:
        """他のモデルと同じバージョンかどうかを比較"""
        if not isinstance(other, BaseModel):
            return NotImplemented
        return self._version == other.version()

File: packages/backend/src/core/domain/baseModel.ts
import { BaseId } from '@shared/utils/core/value-object/base_id';
import { AppBaseEntity } from '../entity/appBaseEntity';
import { Version } from '@shared/utils/features/shared/version';
import { FamilyMemberId } from '@shared/utils/features/family-member/value-object/familyMemberId';
import { ScreenId } from '@shared/utils/features/shared/screenId';
import { RelationValidator } from '@shared/utils/core/validator/relationValidator';

/**
 * ドメインモデルの基底抽象クラス
 * PythonのBaseModelクラスのTypeScript版
 */
export abstract class BaseModel<IdType extends BaseId, EntityType extends AppBaseEntity> {
  protected _id: IdType;
  protected _version: Version;
  protected _createdAt?: Date;
  protected _createdBy?: FamilyMemberId;
  protected _createdFrom?: ScreenId;
  protected _updatedAt?: Date;
  protected _updatedBy?: FamilyMemberId;
  protected _updatedFrom?: ScreenId;
  protected _isUpdated: boolean = false;
  protected _relationValidator: RelationValidator;

  constructor(
    id: IdType,
    version: Version,
    createdAt?: Date,
    createdBy?: FamilyMemberId,
    createdFrom?: ScreenId,
    updatedAt?: Date,
    updatedBy?: FamilyMemberId,
    updatedFrom?: ScreenId
  ) {
    this._id = id;
    this._version = version;
    this._createdAt = createdAt;
    this._createdBy = createdBy;
    this._createdFrom = createdFrom;
    this._updatedAt = updatedAt;
    this._updatedBy = updatedBy;
    this._updatedFrom = updatedFrom;
    this._relationValidator = new RelationValidator();
    this._validate();
  }

  /**
   * モデルの値を検証する（サブクラスで実装必須）
   */
  protected abstract _validate(): void;

  /**
   * IDを取得する
   */
  get id(): IdType {
    return this._id;
  }

  /**
   * バージョンを取得する
   */
  get version(): Version {
    return this._version;
  }

  /**
   * 作成日時を取得する
   */
  get createdAt(): Date | undefined {
    return this._createdAt;
  }

  /**
   * 作成者IDを取得する
   */
  get createdBy(): FamilyMemberId | undefined {
    return this._createdBy;
  }

  /**
   * 作成元スクリーンIDを取得する
   */
  get createdFrom(): ScreenId | undefined {
    return this._createdFrom;
  }

  /**
   * 更新日時を取得する
   */
  get updatedAt(): Date | undefined {
    return this._updatedAt;
  }

  /**
   * 更新者IDを取得する
   */
  get updatedBy(): FamilyMemberId | undefined {
    return this._updatedBy;
  }

  /**
   * 更新元スクリーンIDを取得する
   */
  get updatedFrom(): ScreenId | undefined {
    return this._updatedFrom;
  }

  /**
   * バージョンを1上げ、更新フラグを設定する
   */
  protected _updateVersion(): void {
    if (!this._isUpdated) {
      this._isUpdated = true;
      this._version = new Version(this._version.value + 1);
    }
  }

  /**
   * 他のモデルと同じバージョンかどうかを比較
   */
  isSameVersion(other: BaseModel<any, any>): boolean {
    if (!(other instanceof BaseModel)) {
      return false;
    }
    return this._version.equals(other.version);
  }

  /**
   * ドメインモデルをエンティティに変換する（サブクラスで実装）
   */
  abstract toEntity(): EntityType;
}
